In [82]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from sklearn.datasets import make_moons,make_circles,make_classification
import os
from sklearn.model_selection import train_test_split

In [109]:
np.random.seed(42)

### Downloading the Fashion-MNIST Dataset

In [83]:
import keras 
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.datasets import fashion_mnist

### Data Preprocessing

In [84]:
x_train, y_train, x_valid, y_valid, x_test ,y_test = [], [], [], [], [], []
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()
x_train = x_train.reshape(60000,28,28)
x_test = x_test.reshape(10000,28,28)

In [85]:
x_train, x_valid, y_train, y_valid = train_test_split(x_train, y_train, test_size=0.2)

x_train = x_train.astype('float32')
x_valid = x_valid.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_valid /= 255
x_test /= 255

In [86]:
x_train = x_train.reshape(len(x_train),28,28)
x_test = x_test.reshape(10000,28,28)
x_valid = x_valid.reshape(len(x_valid),28,28)

In [87]:
y_train = keras.utils.to_categorical(y_train, 10)
y_valid = keras.utils.to_categorical(y_valid, 10)
y_test = keras.utils.to_categorical(y_test, 10)

### Building the MLP model

In [88]:
model = Sequential()
model.add(Flatten(input_shape=(28,28)))
model.add(Dense(256, activation='tanh', kernel_initializer = 'he_normal' ,input_shape=(28*28,)))
model.add(Dropout(0.4))
model.add(Dense(128, activation='tanh',kernel_initializer = 'he_normal'))
model.add(Dropout(0.4))
model.add(Dense(100, activation='tanh',kernel_initializer = 'he_normal'))
model.add(Dropout(0.4))
model.add(Dense(10, activation='sigmoid',kernel_initializer = 'he_normal'))
optim = keras.optimizers.SGD(lr=0.01, momentum=0.975, decay=2e-06, nesterov=True)

In [89]:
from timeit import default_timer as timer

class TimingCallback1(keras.callbacks.Callback):
    def __init__(self, logs={}):
        self.logs=[]
    def on_epoch_begin(self, epoch, logs={}):
        self.starttime = timer()
    def on_epoch_end(self, epoch, logs={}):
        self.logs.append(timer()-self.starttime)

In [90]:
model.compile(loss='categorical_crossentropy',optimizer=optim,metrics=['accuracy'])

In [91]:
cb1= TimingCallback1()

In [92]:
history = model.fit(x_train, y_train,batch_size=64,epochs=100,verbose=2,validation_data=(x_valid, y_valid),callbacks=[cb1])

Train on 48000 samples, validate on 12000 samples
Epoch 1/100
 - 3s - loss: 0.8631 - accuracy: 0.6793 - val_loss: 0.5232 - val_accuracy: 0.8149
Epoch 2/100
 - 2s - loss: 0.5965 - accuracy: 0.7931 - val_loss: 0.4786 - val_accuracy: 0.8288
Epoch 3/100
 - 2s - loss: 0.5536 - accuracy: 0.8096 - val_loss: 0.4563 - val_accuracy: 0.8393
Epoch 4/100
 - 2s - loss: 0.5293 - accuracy: 0.8182 - val_loss: 0.4537 - val_accuracy: 0.8399
Epoch 5/100
 - 2s - loss: 0.5138 - accuracy: 0.8233 - val_loss: 0.4559 - val_accuracy: 0.8422
Epoch 6/100
 - 2s - loss: 0.5053 - accuracy: 0.8270 - val_loss: 0.4449 - val_accuracy: 0.8476
Epoch 7/100
 - 2s - loss: 0.4890 - accuracy: 0.8311 - val_loss: 0.4270 - val_accuracy: 0.8431
Epoch 8/100
 - 2s - loss: 0.4754 - accuracy: 0.8362 - val_loss: 0.4064 - val_accuracy: 0.8537
Epoch 9/100
 - 2s - loss: 0.4704 - accuracy: 0.8386 - val_loss: 0.4179 - val_accuracy: 0.8523
Epoch 10/100
 - 2s - loss: 0.4613 - accuracy: 0.8394 - val_loss: 0.4130 - val_accuracy: 0.8574
Epoch 11/

Epoch 87/100
 - 2s - loss: 0.3249 - accuracy: 0.8828 - val_loss: 0.3521 - val_accuracy: 0.8809
Epoch 88/100
 - 2s - loss: 0.3255 - accuracy: 0.8808 - val_loss: 0.3462 - val_accuracy: 0.8795
Epoch 89/100
 - 2s - loss: 0.3253 - accuracy: 0.8832 - val_loss: 0.3511 - val_accuracy: 0.8798
Epoch 90/100
 - 2s - loss: 0.3190 - accuracy: 0.8848 - val_loss: 0.3605 - val_accuracy: 0.8825
Epoch 91/100
 - 2s - loss: 0.3210 - accuracy: 0.8865 - val_loss: 0.3484 - val_accuracy: 0.8787
Epoch 92/100
 - 2s - loss: 0.3225 - accuracy: 0.8853 - val_loss: 0.3543 - val_accuracy: 0.8791
Epoch 93/100
 - 2s - loss: 0.3186 - accuracy: 0.8860 - val_loss: 0.3643 - val_accuracy: 0.8754
Epoch 94/100
 - 2s - loss: 0.3208 - accuracy: 0.8842 - val_loss: 0.3477 - val_accuracy: 0.8773
Epoch 95/100
 - 2s - loss: 0.3197 - accuracy: 0.8844 - val_loss: 0.3563 - val_accuracy: 0.8798
Epoch 96/100
 - 2s - loss: 0.3224 - accuracy: 0.8834 - val_loss: 0.3497 - val_accuracy: 0.8817
Epoch 97/100
 - 2s - loss: 0.3206 - accuracy: 0.88

In [93]:
print("Time taken to train the model: ",sum(cb1.logs))

Time taken to train the model:  219.75140049999573


In [ ]:
#print(model.summary())

In [94]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.38257417073249816
Test accuracy: 0.8708999752998352


### Building the 2nd MLP model varying the number of neurons

In [95]:
model2 = Sequential()
model2.add(Flatten(input_shape=(28,28)))
model2.add(Dense(256, activation='tanh', kernel_initializer = 'he_normal' ,input_shape=(28*28,)))
model2.add(Dropout(0.4))
model2.add(Dense(512, activation='tanh',kernel_initializer = 'he_normal'))
model2.add(Dropout(0.4))
model2.add(Dense(100, activation='tanh',kernel_initializer = 'he_normal'))
model2.add(Dropout(0.4))
model2.add(Dense(10, activation='sigmoid',kernel_initializer = 'he_normal'))
optim = keras.optimizers.SGD(lr=0.01, momentum=0.975, decay=2e-06, nesterov=True)

In [96]:
model2.compile(loss='categorical_crossentropy',optimizer=optim,metrics=['accuracy'])

In [97]:
history = model2.fit(x_train, y_train,batch_size=64,epochs=100,verbose=2,validation_data=(x_valid, y_valid))

Train on 48000 samples, validate on 12000 samples
Epoch 1/100
 - 3s - loss: 0.8301 - accuracy: 0.6900 - val_loss: 0.5279 - val_accuracy: 0.8131
Epoch 2/100
 - 3s - loss: 0.5887 - accuracy: 0.7965 - val_loss: 0.4785 - val_accuracy: 0.8288
Epoch 3/100
 - 3s - loss: 0.5490 - accuracy: 0.8084 - val_loss: 0.4442 - val_accuracy: 0.8444
Epoch 4/100
 - 3s - loss: 0.5238 - accuracy: 0.8182 - val_loss: 0.4265 - val_accuracy: 0.8457
Epoch 5/100
 - 3s - loss: 0.5067 - accuracy: 0.8253 - val_loss: 0.4292 - val_accuracy: 0.8475
Epoch 6/100
 - 3s - loss: 0.4937 - accuracy: 0.8289 - val_loss: 0.4347 - val_accuracy: 0.8475
Epoch 7/100
 - 3s - loss: 0.4822 - accuracy: 0.8314 - val_loss: 0.4115 - val_accuracy: 0.8495
Epoch 8/100
 - 3s - loss: 0.4793 - accuracy: 0.8352 - val_loss: 0.4457 - val_accuracy: 0.8425
Epoch 9/100
 - 3s - loss: 0.4642 - accuracy: 0.8391 - val_loss: 0.4111 - val_accuracy: 0.8527
Epoch 10/100
 - 3s - loss: 0.4568 - accuracy: 0.8400 - val_loss: 0.4123 - val_accuracy: 0.8458
Epoch 11/

Epoch 87/100
 - 3s - loss: 0.3073 - accuracy: 0.8880 - val_loss: 0.3539 - val_accuracy: 0.8771
Epoch 88/100
 - 3s - loss: 0.3073 - accuracy: 0.8895 - val_loss: 0.3478 - val_accuracy: 0.8830
Epoch 89/100
 - 3s - loss: 0.3035 - accuracy: 0.8897 - val_loss: 0.3385 - val_accuracy: 0.8832
Epoch 90/100
 - 3s - loss: 0.3060 - accuracy: 0.8893 - val_loss: 0.3477 - val_accuracy: 0.8802
Epoch 91/100
 - 3s - loss: 0.2997 - accuracy: 0.8901 - val_loss: 0.3506 - val_accuracy: 0.8801
Epoch 92/100
 - 3s - loss: 0.3030 - accuracy: 0.8901 - val_loss: 0.3488 - val_accuracy: 0.8813
Epoch 93/100
 - 3s - loss: 0.3006 - accuracy: 0.8915 - val_loss: 0.3466 - val_accuracy: 0.8806
Epoch 94/100
 - 3s - loss: 0.2979 - accuracy: 0.8922 - val_loss: 0.3478 - val_accuracy: 0.8817
Epoch 95/100
 - 3s - loss: 0.3009 - accuracy: 0.8914 - val_loss: 0.3428 - val_accuracy: 0.8828
Epoch 96/100
 - 3s - loss: 0.2982 - accuracy: 0.8924 - val_loss: 0.3404 - val_accuracy: 0.8832
Epoch 97/100
 - 3s - loss: 0.2972 - accuracy: 0.89

In [ ]:
#print(model2.summary())

In [98]:
score = model2.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy for 2nd model after varying the number of neurons:', score[1])

Test loss: 0.36726623737215996
Test accuracy for 2nd model after varying the number of neurons: 0.8784999847412109


#### After tweaking the number of neurons in the 2nd model:
#### The test accuracy of the classication results is slightly better compared to the 1st model, it has increased from 87.26% to 87.7%.
#### The loss has also decreased from 0.38 to 0.36. Hence, model 2 performs better than the 1st model

### Dimensionality Reduction using PCA

In [99]:
x_train, y_train, x_valid, y_valid, x_test ,y_test = [], [], [], [], [], []
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()
x_train = x_train.reshape(60000,28,28)
x_test = x_test.reshape(10000,28,28)

In [100]:
x_train, x_valid, y_train, y_valid = train_test_split(x_train, y_train, test_size=0.2)

x_train = x_train.astype('float32')
x_valid = x_valid.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_valid /= 255
x_test /= 255

x_train = x_train.reshape(len(x_train),28,28)
x_test = x_test.reshape(10000,28,28)
y_train = keras.utils.to_categorical(y_train, 10)
y_valid = keras.utils.to_categorical(y_valid, 10)
y_test = keras.utils.to_categorical(y_test, 10)

In [101]:
from sklearn.decomposition import PCA

In [102]:
pca = PCA(n_components=0.98)
X_train = x_train.reshape(48000,784)
x_train_reduced = pca.fit_transform(X_train)
print("Explained variance = ",(100*(np.sum(pca.explained_variance_ratio_))))
print("The number of features are reduced to ",pca.n_components_)

Explained variance =  98.00209999084473
The number of features are reduced to  348


In [103]:
X_test = x_test.reshape(10000,784)
x_test_reduced = pca.transform(X_test)
X_valid = x_valid.reshape(12000,784)
x_valid_reduced = pca.transform(X_valid)

### Training a Classifier using the dimensionality reduced dataset

In [104]:
from timeit import default_timer as timer

class TimingCallback(keras.callbacks.Callback):
    def __init__(self, logs={}):
        self.logs=[]
    def on_epoch_begin(self, epoch, logs={}):
        self.starttime = timer()
    def on_epoch_end(self, epoch, logs={}):
        self.logs.append(timer()-self.starttime)

In [105]:
PCAmodel = Sequential()
PCAmodel.add(Dense(256, activation='tanh',input_dim=348,kernel_initializer = 'he_normal'))
PCAmodel.add(Dropout(0.4))
PCAmodel.add(Dense(128, activation='tanh',kernel_initializer = 'he_normal'))
PCAmodel.add(Dropout(0.4))
PCAmodel.add(Dense(100, activation='tanh',kernel_initializer = 'he_normal'))
PCAmodel.add(Dropout(0.4))
PCAmodel.add(Dense(10, activation='sigmoid'))
optim = keras.optimizers.SGD(lr=0.01, momentum=0.975, decay=2e-06, nesterov=True)

In [106]:
PCAmodel.compile(loss='categorical_crossentropy',optimizer=optim,metrics=['accuracy'])
cb = TimingCallback()

In [107]:
history = PCAmodel.fit(x_train_reduced, y_train,batch_size=32,epochs=100,verbose=2,validation_data=(x_valid_reduced,y_valid),callbacks=[cb])

Train on 48000 samples, validate on 12000 samples
Epoch 1/100
 - 4s - loss: 0.8171 - accuracy: 0.7028 - val_loss: 0.4982 - val_accuracy: 0.8236
Epoch 2/100
 - 4s - loss: 0.5900 - accuracy: 0.8029 - val_loss: 0.4671 - val_accuracy: 0.8363
Epoch 3/100
 - 3s - loss: 0.5403 - accuracy: 0.8158 - val_loss: 0.4330 - val_accuracy: 0.8474
Epoch 4/100
 - 4s - loss: 0.5112 - accuracy: 0.8242 - val_loss: 0.4102 - val_accuracy: 0.8553
Epoch 5/100
 - 4s - loss: 0.5005 - accuracy: 0.8302 - val_loss: 0.4153 - val_accuracy: 0.8527
Epoch 6/100
 - 4s - loss: 0.4877 - accuracy: 0.8334 - val_loss: 0.4258 - val_accuracy: 0.8532
Epoch 7/100
 - 4s - loss: 0.4787 - accuracy: 0.8344 - val_loss: 0.4007 - val_accuracy: 0.8597
Epoch 8/100
 - 3s - loss: 0.4695 - accuracy: 0.8369 - val_loss: 0.4278 - val_accuracy: 0.8531
Epoch 9/100
 - 3s - loss: 0.4598 - accuracy: 0.8410 - val_loss: 0.3957 - val_accuracy: 0.8616
Epoch 10/100
 - 3s - loss: 0.4514 - accuracy: 0.8451 - val_loss: 0.3891 - val_accuracy: 0.8605
Epoch 11/

Epoch 87/100
 - 3s - loss: 0.3289 - accuracy: 0.8819 - val_loss: 0.3534 - val_accuracy: 0.8807
Epoch 88/100
 - 3s - loss: 0.3305 - accuracy: 0.8824 - val_loss: 0.3378 - val_accuracy: 0.8816
Epoch 89/100
 - 3s - loss: 0.3260 - accuracy: 0.8828 - val_loss: 0.3431 - val_accuracy: 0.8820
Epoch 90/100
 - 3s - loss: 0.3242 - accuracy: 0.8810 - val_loss: 0.3440 - val_accuracy: 0.8880
Epoch 91/100
 - 3s - loss: 0.3241 - accuracy: 0.8837 - val_loss: 0.3456 - val_accuracy: 0.8892
Epoch 92/100
 - 3s - loss: 0.3237 - accuracy: 0.8825 - val_loss: 0.3429 - val_accuracy: 0.8883
Epoch 93/100
 - 3s - loss: 0.3205 - accuracy: 0.8840 - val_loss: 0.3406 - val_accuracy: 0.8875
Epoch 94/100
 - 3s - loss: 0.3184 - accuracy: 0.8860 - val_loss: 0.3413 - val_accuracy: 0.8848
Epoch 95/100
 - 3s - loss: 0.3177 - accuracy: 0.8840 - val_loss: 0.3505 - val_accuracy: 0.8858
Epoch 96/100
 - 3s - loss: 0.3213 - accuracy: 0.8828 - val_loss: 0.3407 - val_accuracy: 0.8821
Epoch 97/100
 - 3s - loss: 0.3144 - accuracy: 0.88

In [108]:
print(sum(cb.logs))

348.32934510000086


In [ ]:
#PCAmodel.summary()

In [ ]:
score = PCAmodel.evaluate(x_test_reduced, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

#### Classification Accuracy: The test accuracy is better on the dataset with PCA applied with 88% compared to 87.2% on the original dataset.
#### --
#### Training Time: The PCA model which has been trained on almost half of the original features took approximately 323 seconds or 5 minutes whereas the model on the original dataset took 219 seconds or 3.65 minutes.
#### The training was faster on the original dataset using the same number of layers, neurons and parameters.

### RandomizedSearchCV

In [40]:
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import RandomizedSearchCV

In [72]:
def build_model(var_activation='tanh',var_optimizer='SGD'):
    model = Sequential()
    model.add(Dense(256,activation=var_activation,input_dim=348))
    model.add(Dropout(0.4))
    model.add(Dense(128,activation=var_activation))
    model.add(Dropout(0.4))
    model.add(Dense(100,activation=var_activation))
    model.add(Dropout(0.4))
    model.add(Dense(10,activation='sigmoid'))
    model.compile(loss='categorical_crossentropy',
                optimizer=var_optimizer,
                metrics=["accuracy"])
    return model

In [73]:
model_default = build_model()
model_default.summary()

Model: "sequential_83"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_307 (Dense)            (None, 256)               89344     
_________________________________________________________________
dropout_227 (Dropout)        (None, 256)               0         
_________________________________________________________________
dense_308 (Dense)            (None, 128)               32896     
_________________________________________________________________
dropout_228 (Dropout)        (None, 128)               0         
_________________________________________________________________
dense_309 (Dense)            (None, 100)               12900     
_________________________________________________________________
dropout_229 (Dropout)        (None, 100)               0         
_________________________________________________________________
dense_310 (Dense)            (None, 10)              

In [74]:
_activations=['tanh','relu','selu','leaky_relu(alpha=0.01)']
_optimizers=['sgd','adam']
_batch_size=[16,32,64]
params=dict(var_activation=_activations,
            var_optimizer=_optimizers,
            batch_size=_batch_size,
           )
print(params)

{'var_activation': ['tanh', 'relu', 'selu', 'leaky_relu(alpha=0.01)'], 'var_optimizer': ['sgd', 'adam'], 'batch_size': [16, 32, 64]}


In [75]:
model = KerasClassifier(build_fn=build_model,epochs=50,batch_size=32)

In [76]:
rscv = RandomizedSearchCV(model, param_distributions=params, cv=3, n_iter=10)
rscv_results = rscv.fit(x_train_reduced,y_train)

Epoch 1/50
32000/32000 [==============================] - 4s 128us/step - loss: 0.7211 - accuracy: 0.7582
Epoch 2/50
32000/32000 [==============================] - 4s 120us/step - loss: 0.4944 - accuracy: 0.8286
Epoch 3/50
32000/32000 [==============================] - 4s 118us/step - loss: 0.4596 - accuracy: 0.8392
Epoch 4/50
32000/32000 [==============================] - 4s 118us/step - loss: 0.4333 - accuracy: 0.8467
Epoch 5/50
32000/32000 [==============================] - 4s 118us/step - loss: 0.4204 - accuracy: 0.8509
Epoch 6/50
32000/32000 [==============================] - 4s 119us/step - loss: 0.4042 - accuracy: 0.8572
Epoch 7/50
32000/32000 [==============================] - 4s 119us/step - loss: 0.3975 - accuracy: 0.8600
Epoch 8/50
32000/32000 [==============================] - 4s 118us/step - loss: 0.3913 - accuracy: 0.8618
Epoch 9/50
32000/32000 [==============================] - 4s 119us/step - loss: 0.3777 - accuracy: 0.8655
Epoch 10/50
32000/32000 [=====================

32000/32000 [==============================] - 4s 114us/step - loss: 0.3058 - accuracy: 0.8918
Epoch 28/50
32000/32000 [==============================] - 4s 118us/step - loss: 0.2984 - accuracy: 0.8942
Epoch 29/50
32000/32000 [==============================] - 4s 117us/step - loss: 0.2976 - accuracy: 0.8934
Epoch 30/50
32000/32000 [==============================] - 4s 118us/step - loss: 0.2897 - accuracy: 0.8963
Epoch 31/50
32000/32000 [==============================] - 4s 119us/step - loss: 0.2880 - accuracy: 0.8979
Epoch 32/50
32000/32000 [==============================] - 4s 123us/step - loss: 0.2903 - accuracy: 0.8985
Epoch 33/50
32000/32000 [==============================] - 4s 119us/step - loss: 0.2884 - accuracy: 0.8974
Epoch 34/50
32000/32000 [==============================] - 4s 112us/step - loss: 0.2817 - accuracy: 0.9003
Epoch 35/50
32000/32000 [==============================] - 4s 115us/step - loss: 0.2825 - accuracy: 0.8995
Epoch 36/50
32000/32000 [========================

32000/32000 [==============================] - 2s 73us/step - loss: 0.4943 - accuracy: 0.8283
Epoch 3/50
32000/32000 [==============================] - 2s 72us/step - loss: 0.4603 - accuracy: 0.8418
Epoch 4/50
32000/32000 [==============================] - 2s 66us/step - loss: 0.4384 - accuracy: 0.8466
Epoch 5/50
32000/32000 [==============================] - 2s 67us/step - loss: 0.4216 - accuracy: 0.8549
Epoch 6/50
32000/32000 [==============================] - 2s 67us/step - loss: 0.4077 - accuracy: 0.8556
Epoch 7/50
32000/32000 [==============================] - 2s 67us/step - loss: 0.3991 - accuracy: 0.8584
Epoch 8/50
32000/32000 [==============================] - 2s 69us/step - loss: 0.3897 - accuracy: 0.8633
Epoch 9/50
32000/32000 [==============================] - 2s 71us/step - loss: 0.3812 - accuracy: 0.8659
Epoch 10/50
32000/32000 [==============================] - 2s 70us/step - loss: 0.3755 - accuracy: 0.8665
Epoch 11/50
32000/32000 [==============================] - 2s 70u

Epoch 29/50
32000/32000 [==============================] - 2s 72us/step - loss: 0.2994 - accuracy: 0.8946
Epoch 30/50
32000/32000 [==============================] - 2s 72us/step - loss: 0.2992 - accuracy: 0.8912
Epoch 31/50
32000/32000 [==============================] - 2s 72us/step - loss: 0.2956 - accuracy: 0.8942
Epoch 32/50
32000/32000 [==============================] - 2s 73us/step - loss: 0.2935 - accuracy: 0.8945
Epoch 33/50
32000/32000 [==============================] - 2s 72us/step - loss: 0.2893 - accuracy: 0.8964
Epoch 34/50
32000/32000 [==============================] - 2s 72us/step - loss: 0.2845 - accuracy: 0.8968
Epoch 35/50
32000/32000 [==============================] - 2s 71us/step - loss: 0.2855 - accuracy: 0.8965
Epoch 36/50
32000/32000 [==============================] - 2s 73us/step - loss: 0.2834 - accuracy: 0.8981
Epoch 37/50
32000/32000 [==============================] - 2s 70us/step - loss: 0.2835 - accuracy: 0.8991
Epoch 38/50
32000/32000 [=====================

C:\Users\Sudipto\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: Unknown activation function:leaky_relu(alpha=0.01)

  FitFailedWarning)


Epoch 1/50
32000/32000 [==============================] - 2s 67us/step - loss: 1.7078 - accuracy: 0.4816
Epoch 2/50
32000/32000 [==============================] - 2s 63us/step - loss: 1.2530 - accuracy: 0.6062
Epoch 3/50
32000/32000 [==============================] - 2s 64us/step - loss: 0.8758 - accuracy: 0.6978
Epoch 4/50
32000/32000 [==============================] - 2s 64us/step - loss: 0.7310 - accuracy: 0.7442
Epoch 5/50
32000/32000 [==============================] - 2s 64us/step - loss: 0.6613 - accuracy: 0.7669
Epoch 6/50
32000/32000 [==============================] - 2s 63us/step - loss: 0.6236 - accuracy: 0.7832
Epoch 7/50
32000/32000 [==============================] - 2s 63us/step - loss: 0.5879 - accuracy: 0.7961
Epoch 8/50
32000/32000 [==============================] - 2s 62us/step - loss: 0.5713 - accuracy: 0.8003
Epoch 9/50
32000/32000 [==============================] - 2s 63us/step - loss: 0.5511 - accuracy: 0.8069
Epoch 10/50
32000/32000 [==============================

32000/32000 [==============================] - 2s 60us/step - loss: 0.4445 - accuracy: 0.8423
Epoch 28/50
32000/32000 [==============================] - 2s 64us/step - loss: 0.4408 - accuracy: 0.8439
Epoch 29/50
32000/32000 [==============================] - 2s 63us/step - loss: 0.4391 - accuracy: 0.8456
Epoch 30/50
32000/32000 [==============================] - 2s 63us/step - loss: 0.4345 - accuracy: 0.8462
Epoch 31/50
32000/32000 [==============================] - 2s 63us/step - loss: 0.4357 - accuracy: 0.8485
Epoch 32/50
32000/32000 [==============================] - 2s 60us/step - loss: 0.4315 - accuracy: 0.8481
Epoch 33/50
32000/32000 [==============================] - 2s 60us/step - loss: 0.4308 - accuracy: 0.8492
Epoch 34/50
32000/32000 [==============================] - 2s 60us/step - loss: 0.4307 - accuracy: 0.8467
Epoch 35/50
32000/32000 [==============================] - 2s 57us/step - loss: 0.4303 - accuracy: 0.8474
Epoch 36/50
32000/32000 [==============================] -

32000/32000 [==============================] - 1s 38us/step - loss: 1.4052 - accuracy: 0.5409
Epoch 4/50
32000/32000 [==============================] - 1s 38us/step - loss: 1.1974 - accuracy: 0.5838
Epoch 5/50
32000/32000 [==============================] - 1s 38us/step - loss: 0.9758 - accuracy: 0.6669
Epoch 6/50
32000/32000 [==============================] - 1s 38us/step - loss: 0.8327 - accuracy: 0.7105
Epoch 7/50
32000/32000 [==============================] - 1s 37us/step - loss: 0.7607 - accuracy: 0.7320
Epoch 8/50
32000/32000 [==============================] - 1s 38us/step - loss: 0.7132 - accuracy: 0.7479
Epoch 9/50
32000/32000 [==============================] - 1s 37us/step - loss: 0.6770 - accuracy: 0.7600
Epoch 10/50
32000/32000 [==============================] - 1s 38us/step - loss: 0.6498 - accuracy: 0.7741
Epoch 11/50
32000/32000 [==============================] - 1s 37us/step - loss: 0.6240 - accuracy: 0.7786
Epoch 12/50
32000/32000 [==============================] - 1s 38

Epoch 30/50
32000/32000 [==============================] - 1s 35us/step - loss: 0.4786 - accuracy: 0.8321
Epoch 31/50
32000/32000 [==============================] - 1s 35us/step - loss: 0.4837 - accuracy: 0.8316
Epoch 32/50
32000/32000 [==============================] - 1s 35us/step - loss: 0.4770 - accuracy: 0.8357
Epoch 33/50
32000/32000 [==============================] - 1s 35us/step - loss: 0.4759 - accuracy: 0.8338
Epoch 34/50
32000/32000 [==============================] - 1s 35us/step - loss: 0.4752 - accuracy: 0.8372
Epoch 35/50
32000/32000 [==============================] - 1s 35us/step - loss: 0.4725 - accuracy: 0.8372
Epoch 36/50
32000/32000 [==============================] - 1s 35us/step - loss: 0.4672 - accuracy: 0.8351
Epoch 37/50
32000/32000 [==============================] - 1s 35us/step - loss: 0.4689 - accuracy: 0.8364
Epoch 38/50
32000/32000 [==============================] - 1s 34us/step - loss: 0.4607 - accuracy: 0.8407
Epoch 39/50
32000/32000 [=====================

32000/32000 [==============================] - 4s 120us/step - loss: 0.5366 - accuracy: 0.8125
Epoch 7/50
32000/32000 [==============================] - 4s 116us/step - loss: 0.5129 - accuracy: 0.8217
Epoch 8/50
32000/32000 [==============================] - 4s 112us/step - loss: 0.5071 - accuracy: 0.8210
Epoch 9/50
32000/32000 [==============================] - 4s 120us/step - loss: 0.4990 - accuracy: 0.8258
Epoch 10/50
32000/32000 [==============================] - 4s 113us/step - loss: 0.4915 - accuracy: 0.8288
Epoch 11/50
32000/32000 [==============================] - 4s 113us/step - loss: 0.4815 - accuracy: 0.8301
Epoch 12/50
32000/32000 [==============================] - 4s 113us/step - loss: 0.4819 - accuracy: 0.8280
Epoch 13/50
32000/32000 [==============================] - 4s 112us/step - loss: 0.4765 - accuracy: 0.8361
Epoch 14/50
32000/32000 [==============================] - 4s 112us/step - loss: 0.4660 - accuracy: 0.8347
Epoch 15/50
32000/32000 [===========================

32000/32000 [==============================] - 4s 113us/step - loss: 0.4202 - accuracy: 0.8516
Epoch 33/50
32000/32000 [==============================] - 4s 114us/step - loss: 0.4214 - accuracy: 0.8486
Epoch 34/50
32000/32000 [==============================] - 4s 113us/step - loss: 0.4150 - accuracy: 0.8541
Epoch 35/50
32000/32000 [==============================] - 4s 112us/step - loss: 0.4154 - accuracy: 0.8537
Epoch 36/50
32000/32000 [==============================] - 4s 114us/step - loss: 0.4092 - accuracy: 0.8547
Epoch 37/50
32000/32000 [==============================] - 3s 109us/step - loss: 0.4107 - accuracy: 0.8551
Epoch 38/50
32000/32000 [==============================] - 3s 106us/step - loss: 0.4123 - accuracy: 0.8529
Epoch 39/50
32000/32000 [==============================] - 3s 105us/step - loss: 0.4094 - accuracy: 0.8528
Epoch 40/50
32000/32000 [==============================] - 4s 111us/step - loss: 0.4126 - accuracy: 0.8528
Epoch 41/50
32000/32000 [========================

32000/32000 [==============================] - 1s 40us/step - loss: 0.3138 - accuracy: 0.8874
Epoch 8/50
32000/32000 [==============================] - 1s 40us/step - loss: 0.2965 - accuracy: 0.8944
Epoch 9/50
32000/32000 [==============================] - 1s 40us/step - loss: 0.2829 - accuracy: 0.8984
Epoch 10/50
32000/32000 [==============================] - 1s 40us/step - loss: 0.2641 - accuracy: 0.9048
Epoch 11/50
32000/32000 [==============================] - 1s 40us/step - loss: 0.2559 - accuracy: 0.9063
Epoch 12/50
32000/32000 [==============================] - 1s 42us/step - loss: 0.2442 - accuracy: 0.9116
Epoch 13/50
32000/32000 [==============================] - 1s 41us/step - loss: 0.2308 - accuracy: 0.9158
Epoch 14/50
32000/32000 [==============================] - 1s 42us/step - loss: 0.2245 - accuracy: 0.9188
Epoch 15/50
32000/32000 [==============================] - 1s 43us/step - loss: 0.2153 - accuracy: 0.9209
Epoch 16/50
32000/32000 [==============================] - 1

32000/32000 [==============================] - 1s 41us/step - loss: 0.1236 - accuracy: 0.9564
Epoch 34/50
32000/32000 [==============================] - 1s 41us/step - loss: 0.1204 - accuracy: 0.9560
Epoch 35/50
32000/32000 [==============================] - 1s 43us/step - loss: 0.1232 - accuracy: 0.9563
Epoch 36/50
32000/32000 [==============================] - 1s 43us/step - loss: 0.1153 - accuracy: 0.9587
Epoch 37/50
32000/32000 [==============================] - 1s 43us/step - loss: 0.1172 - accuracy: 0.9592
Epoch 38/50
32000/32000 [==============================] - 1s 43us/step - loss: 0.1108 - accuracy: 0.9607
Epoch 39/50
32000/32000 [==============================] - 1s 43us/step - loss: 0.1140 - accuracy: 0.9603
Epoch 40/50
32000/32000 [==============================] - 1s 43us/step - loss: 0.1126 - accuracy: 0.9611
Epoch 41/50
32000/32000 [==============================] - 1s 43us/step - loss: 0.1055 - accuracy: 0.9634
Epoch 42/50
32000/32000 [==============================] -

32000/32000 [==============================] - 3s 85us/step - loss: 0.4887 - accuracy: 0.8292
Epoch 10/50
32000/32000 [==============================] - 3s 86us/step - loss: 0.4764 - accuracy: 0.8334
Epoch 11/50
32000/32000 [==============================] - 3s 87us/step - loss: 0.4719 - accuracy: 0.8353
Epoch 12/50
32000/32000 [==============================] - 3s 86us/step - loss: 0.4670 - accuracy: 0.8335
Epoch 13/50
32000/32000 [==============================] - 3s 86us/step - loss: 0.4558 - accuracy: 0.8410
Epoch 14/50
32000/32000 [==============================] - 3s 85us/step - loss: 0.4520 - accuracy: 0.8422
Epoch 15/50
32000/32000 [==============================] - 3s 86us/step - loss: 0.4490 - accuracy: 0.8426
Epoch 16/50
32000/32000 [==============================] - 3s 87us/step - loss: 0.4425 - accuracy: 0.8441
Epoch 17/50
32000/32000 [==============================] - 3s 86us/step - loss: 0.4392 - accuracy: 0.8461
Epoch 18/50
32000/32000 [==============================] -

32000/32000 [==============================] - 3s 87us/step - loss: 0.3923 - accuracy: 0.8622
Epoch 36/50
32000/32000 [==============================] - 3s 86us/step - loss: 0.3912 - accuracy: 0.8619
Epoch 37/50
32000/32000 [==============================] - 3s 88us/step - loss: 0.3899 - accuracy: 0.8613
Epoch 38/50
32000/32000 [==============================] - 3s 86us/step - loss: 0.3888 - accuracy: 0.8616
Epoch 39/50
32000/32000 [==============================] - 3s 88us/step - loss: 0.3883 - accuracy: 0.8631
Epoch 40/50
32000/32000 [==============================] - 3s 93us/step - loss: 0.3877 - accuracy: 0.8613
Epoch 41/50
32000/32000 [==============================] - 3s 88us/step - loss: 0.3867 - accuracy: 0.8630
Epoch 42/50
32000/32000 [==============================] - 3s 87us/step - loss: 0.3848 - accuracy: 0.8629
Epoch 43/50
32000/32000 [==============================] - 3s 86us/step - loss: 0.3834 - accuracy: 0.8624
Epoch 44/50
32000/32000 [==============================] -

32000/32000 [==============================] - 1s 39us/step - loss: 0.3731 - accuracy: 0.8668
Epoch 12/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.3693 - accuracy: 0.8678
Epoch 13/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.3590 - accuracy: 0.8694
Epoch 14/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.3591 - accuracy: 0.8712
Epoch 15/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.3547 - accuracy: 0.8730
Epoch 16/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.3536 - accuracy: 0.8737
Epoch 17/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.3468 - accuracy: 0.8770
Epoch 18/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.3426 - accuracy: 0.8757
Epoch 19/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.3375 - accuracy: 0.8771
Epoch 20/50
32000/32000 [==============================] -

32000/32000 [==============================] - 1s 40us/step - loss: 0.2925 - accuracy: 0.8947
Epoch 38/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.2876 - accuracy: 0.8970
Epoch 39/50
32000/32000 [==============================] - 1s 40us/step - loss: 0.2880 - accuracy: 0.8948
Epoch 40/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.2878 - accuracy: 0.8952
Epoch 41/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.2812 - accuracy: 0.8972
Epoch 42/50
32000/32000 [==============================] - 1s 39us/step - loss: 0.2791 - accuracy: 0.8990
Epoch 43/50
32000/32000 [==============================] - 1s 40us/step - loss: 0.2747 - accuracy: 0.8998
Epoch 44/50
32000/32000 [==============================] - 1s 41us/step - loss: 0.2768 - accuracy: 0.8999
Epoch 45/50
32000/32000 [==============================] - 1s 41us/step - loss: 0.2742 - accuracy: 0.9019
Epoch 46/50
32000/32000 [==============================] -

48000/48000 [==============================] - 2s 46us/step - loss: 0.2314 - accuracy: 0.9177
Epoch 14/50
48000/48000 [==============================] - 2s 46us/step - loss: 0.2239 - accuracy: 0.9185
Epoch 15/50
48000/48000 [==============================] - 2s 46us/step - loss: 0.2193 - accuracy: 0.9202
Epoch 16/50
48000/48000 [==============================] - 2s 46us/step - loss: 0.2130 - accuracy: 0.9229
Epoch 17/50
48000/48000 [==============================] - 2s 46us/step - loss: 0.2069 - accuracy: 0.9251
Epoch 18/50
48000/48000 [==============================] - 2s 46us/step - loss: 0.2013 - accuracy: 0.9264
Epoch 19/50
48000/48000 [==============================] - 2s 47us/step - loss: 0.1973 - accuracy: 0.9292
Epoch 20/50
48000/48000 [==============================] - 2s 47us/step - loss: 0.1924 - accuracy: 0.9305
Epoch 21/50
48000/48000 [==============================] - 2s 47us/step - loss: 0.1878 - accuracy: 0.9310
Epoch 22/50
48000/48000 [==============================] -

In [77]:
print('Best score is: {} using {}'.format(rscv_results.best_score_,
rscv_results.best_params_))

Best score is: 0.8896458347638448 using {'var_optimizer': 'adam', 'var_activation': 'relu', 'batch_size': 64}


In [78]:
pred = rscv.predict(x_test_reduced)

In [81]:
score = rscv.score(x_test_reduced,y_test)
score

10000/10000 [==============================] - 0s 13us/step


0.8902000188827515

### After tuning Hyperparameters using RandomisedSearchCV, the test accuracy is 89% for 50 epochs compared to 88% accuracy on the dataset after applying PCA. The accuracy would increase if the number of epochs are increased to 100 but the training time would increase significantly.